File: app/src/main/java/at/shockbytes/corey/ui/fragment/CreateWorkoutFragment.java
package at.shockbytes.corey.ui.fragment;


import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import at.shockbytes.corey.R;
import at.shockbytes.corey.adapter.ExerciseAdapter;
import at.shockbytes.corey.adapter.WorkoutCraftingSpinnerAdapter;
import at.shockbytes.corey.common.core.workout.model.Exercise;
import at.shockbytes.corey.common.core.workout.model.Workout;
import at.shockbytes.corey.dagger.AppComponent;
import at.shockbytes.corey.ui.fragment.dialogs.AddExercisesDialogFragment;
import at.shockbytes.corey.util.AppCoreyUtils;
import at.shockbytes.corey.util.AppParams;
import at.shockbytes.util.adapter.BaseItemTouchHelper;
import at.shockbytes.util.view.ViewManager;
import butterknife.BindView;
import butterknife.OnClick;

@SuppressWarnings("ConstantConditions")
public class CreateWorkoutFragment extends BaseFragment implements AdapterView.OnItemSelectedListener,
        AddExercisesDialogFragment.OnExerciseCreatedListener {

    private enum CardState {EXPANDED, COLLAPSED}

    public interface OnChangeSystemBarsColorListener {

        void onChangeSystemBarsColor(int from, int fromDark, int to, int toDark);
    }

    private static final String ARG_WORKOUT = "arg_workout_edit";

    public static CreateWorkoutFragment newInstance(Workout workout) {
        CreateWorkoutFragment fragment = new CreateWorkoutFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_WORKOUT, workout);
        fragment.setArguments(args);
        return fragment;
    }

    private Workout workout;
    private boolean isUpdateMode;

    private CardState cardState;
    private OnChangeSystemBarsColorListener listener;

    private int[] oldColors;
    private final int allColors[][] = new int[][] {
            {R.color.workout_intensity_easy, R.color.workout_intensity_easy_dark},
            {R.color.workout_intensity_medium, R.color.workout_intensity_medium_dark},
            {R.color.workout_intensity_hard, R.color.workout_intensity_hard_dark},
            {R.color.workout_intensity_beast, R.color.workout_intensity_beast_dark}
    };


    private ExerciseAdapter exerciseAdapter;

    @BindView(R.id.fragment_create_workout_edit_name)
    protected EditText editName;

    @BindView(R.id.fragment_create_workout_edit_duration)
    protected EditText editDuration;

    @BindView(R.id.fragment_create_workout_spinner_body_region)
    protected Spinner spinnerBodyRegion;

    @BindView(R.id.fragment_create_workout_spinner_intensity)
    protected Spinner spinnerIntensity;

    @BindView(R.id.fragment_create_workout_recyclerview_exercises)
    protected RecyclerView recyclerViewExercises;

    @BindView(R.id.fragment_create_workout_btn_exp_col_general)
    protected ImageButton imgBtnExpandCollapse;

    @BindView(R.id.fragment_create_workout_general_collapse_container)
    protected View generalCollapseContainer;

    public CreateWorkoutFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        oldColors = new int[] {R.color.workout_intensity_medium, R.color.workout_intensity_medium_dark};

        cardState = CardState.EXPANDED;
        Workout updated = getArguments().getParcelable(ARG_WORKOUT);
        isUpdateMode = updated != null;
        if (updated != null) {
            workout = updated;
        } else {
            workout = new Workout();
        }
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_create_workout;
    }

    @Override
    protected void injectToGraph(@NotNull AppComponent appComponent) {
        // Do nothing
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_create_workout, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == R.id.menu_create_workout_done) {
            validateInput();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
            listener = (OnChangeSystemBarsColorListener) context;
        } catch (IllegalStateException e) {
            Log.e("Corey", "Attached activity must implement listener!");
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {

        if (i > 0 && listener != null) {
            int[] newColors = allColors[i-1].clone();
            listener.onChangeSystemBarsColor(oldColors[0], oldColors[1], newColors[0], newColors[1]);
            oldColors = newColors.clone();
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {
    }

    @Override
    public void onExerciseCreated(Exercise exercise) {
        //exercise.normalizeNameForJson();
        exerciseAdapter.addEntityAtLast(exercise);
    }

    @OnClick(R.id.fragment_create_workout_btn_exp_col_general)
    protected void onClickExpandCollapse() {

        if (cardState == CardState.EXPANDED) {
            ViewManager.INSTANCE.collapse(generalCollapseContainer);
            imgBtnExpandCollapse.animate().rotation(-180).start();
            cardState = CardState.COLLAPSED;
        } else if (cardState == CardState.COLLAPSED) {
            ViewManager.INSTANCE.expand(generalCollapseContainer);
            imgBtnExpandCollapse.animate().rotation(0).start();
            cardState = CardState.EXPANDED;
        }
    }

    @OnClick(R.id.fragment_create_workout_btn_add_exercise)
    protected void onClickAddExercise() {

        if (cardState == CardState.EXPANDED) {
            onClickExpandCollapse();
        }

        AddExercisesDialogFragment fragment = AddExercisesDialogFragment.newInstance();
        fragment.setOnExerciseCreatedListener(this);
        fragment.show(getFragmentManager(), fragment.getTag());
    }

    @Override
    public void setupViews() {

        spinnerBodyRegion.setAdapter(new WorkoutCraftingSpinnerAdapter(getContext(),
                AppCoreyUtils.INSTANCE.getBodyRegionSpinnerData(getContext())));
        spinnerIntensity.setAdapter(new WorkoutCraftingSpinnerAdapter(getContext(),
                AppCoreyUtils.INSTANCE.getIntensitySpinnerData(getContext())));
        spinnerIntensity.setOnItemSelectedListener(this);

        recyclerViewExercises.setLayoutManager(new LinearLayoutManager(getContext()));
        exerciseAdapter = new ExerciseAdapter(getContext(), new ArrayList<Exercise>());
        exerciseAdapter.setItemsMovable(true);
        //exerciseAdapter.setOnItemMoveListener(this);
        ItemTouchHelper.Callback callback = new BaseItemTouchHelper(exerciseAdapter, true, BaseItemTouchHelper.DragAccess.VERTICAL);
        ItemTouchHelper touchHelper = new ItemTouchHelper(callback);
        touchHelper.attachToRecyclerView(recyclerViewExercises);
        recyclerViewExercises.setAdapter(exerciseAdapter);

        if (isUpdateMode) {
            fillFields();
        }
    }

    private void validateInput() {

        // Name -- Not empty
        String name = editName.getText().toString();
        if (name.isEmpty()) {
            Snackbar.make(getView(), R.string.validation_empty_name, Snackbar.LENGTH_LONG).show();
            return;
        }

        // Duration -- Not empty
        String strDuration = editDuration.getText().toString();
        if (strDuration.isEmpty()) {
            Snackbar.make(getView(), R.string.validation_empty_duration, Snackbar.LENGTH_LONG).show();
            return;
        }
        int duration = Integer.parseInt(strDuration);

        // Body Region -- Not first selected
        int brIdx = spinnerBodyRegion.getSelectedItemPosition();
        if (brIdx <= 0) {
            Snackbar.make(getView(), R.string.validation_empty_body_region, Snackbar.LENGTH_LONG).show();
            return;
        }

        // Intensity -- Not first selected
        int inIdx = spinnerIntensity.getSelectedItemPosition();
        if (brIdx <= 0) {
            Snackbar.make(getView(), R.string.validation_empty_intensity, Snackbar.LENGTH_LONG).show();
            return;
        }

        // Workout items -- Not empty
        List<Exercise> exercises = exerciseAdapter.getData();
        if (exercises.size() == 0) {
            Snackbar.make(getView(), R.string.validation_empty_exercises, Snackbar.LENGTH_LONG).show();
            return;
        }

        workout.setName(name);
        workout.setDuration(duration);
        workout.setBodyRegion(Workout.BodyRegion.values()[brIdx-1]);
        workout.setIntensity(Workout.Intensity.values()[inIdx-1]);
        workout.setExercises(exercises);

        notifyMainActivity();
    }


    private void fillFields() {

        editName.setText(workout.getDisplayableName());
        editDuration.setText(String.valueOf(workout.getDuration()));
        int brIdx = workout.getBodyRegion().ordinal() + 1;
        spinnerBodyRegion.setSelection(brIdx, true);
        int inIdx = workout.getIntensity().ordinal() + 1;
        spinnerIntensity.setSelection(inIdx, true);
        exerciseAdapter.setData(workout.getExercises());
    }

    private void notifyMainActivity() {

        Intent data = new Intent()
                .putExtra(AppParams.INTENT_EXTRA_NEW_WORKOUT, workout)
                .putExtra(AppParams.INTENT_EXTRA_WORKOUT_UPDATED, isUpdateMode);
        getActivity().setResult(Activity.RESULT_OK, data);
        getActivity().supportFinishAfterTransition();
    }


}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/CreateWorkoutFragment.kt
package at.shockbytes.corey.ui.fragment

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.ItemTouchHelper
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Spinner
import at.shockbytes.core.ui.activity.base.TintableBackNavigableActivity
import at.shockbytes.core.ui.fragment.BaseFragment
import at.shockbytes.corey.R
import at.shockbytes.corey.ui.adapter.ExerciseAdapter
import at.shockbytes.corey.ui.adapter.spinner.WorkoutCraftingSpinnerAdapter
import at.shockbytes.corey.common.core.workout.model.Workout
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.ui.fragment.dialog.AddExercisesDialogFragment
import at.shockbytes.corey.util.CoreyAppUtils
import at.shockbytes.corey.util.AppParams
import at.shockbytes.util.adapter.BaseItemTouchHelper
import at.shockbytes.util.view.ViewManager
import kotlinx.android.synthetic.main.fragment_create_workout.*
import kotterknife.bindView

class CreateWorkoutFragment : BaseFragment<AppComponent>(), AdapterView.OnItemSelectedListener {

    override val snackBarBackgroundColorRes: Int = R.color.sb_background
    override val snackBarForegroundColorRes: Int = R.color.sb_foreground

    private enum class CardState {
        EXPANDED, COLLAPSED
    }

    private var workout: Workout = Workout() // Default initialize it to avoid nullability
    private var isUpdateMode: Boolean = false
    private var cardState: CardState = CardState.EXPANDED

    private val allColors = arrayOf(
            intArrayOf(R.color.workout_intensity_easy, R.color.workout_intensity_easy_dark),
            intArrayOf(R.color.workout_intensity_medium, R.color.workout_intensity_medium_dark),
            intArrayOf(R.color.workout_intensity_hard, R.color.workout_intensity_hard_dark),
            intArrayOf(R.color.workout_intensity_beast, R.color.workout_intensity_beast_dark))

    private lateinit var exerciseAdapter: ExerciseAdapter

    private val editName: EditText by bindView(R.id.fragment_create_workout_edit_name)
    private val editDuration: EditText by bindView(R.id.fragment_create_workout_edit_duration)
    private val spinnerBodyRegion: Spinner by bindView(R.id.fragment_create_workout_spinner_body_region)
    private val spinnerIntensity: Spinner by bindView(R.id.fragment_create_workout_spinner_intensity)
    private val recyclerViewExercises: androidx.recyclerview.widget.RecyclerView by bindView(R.id.fragment_create_workout_recyclerview_exercises)
    private val imgBtnExpandCollapse: ImageButton by bindView(R.id.fragment_create_workout_btn_exp_col_general)
    private val generalCollapseContainer: View by bindView(R.id.fragment_create_workout_general_collapse_container)

    override val layoutId = R.layout.fragment_create_workout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        val updated = arguments?.getParcelable<Workout>(ARG_WORKOUT)
        isUpdateMode = updated != null
        if (updated != null) {
            workout = updated
        }
    }

    override fun injectToGraph(appComponent: AppComponent?) = Unit

    override fun bindViewModel() = Unit

    override fun unbindViewModel() = Unit

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_create_workout, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_create_workout_done) {
            validateInput()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        if (i > 0) {
            val newColors = allColors[i - 1].clone()

            (activity as? TintableBackNavigableActivity<*>)
                    ?.tintSystemBarsWithText(
                            actionBarTextColor = Color.WHITE,
                            actionBarColor = ContextCompat.getColor(context!!, newColors[0]),
                            statusBarColor = ContextCompat.getColor(context!!, newColors[1]),
                            title = null,
                            animated = true,
                            useSameColorsForBoth = false
                    )
        }
    }

    override fun onNothingSelected(adapterView: AdapterView<*>) {}

    override fun setupViews() {

        context?.let { ctx ->

            spinnerBodyRegion.adapter = WorkoutCraftingSpinnerAdapter(ctx,
                CoreyAppUtils.getBodyRegionSpinnerData(ctx))
            spinnerIntensity.adapter = WorkoutCraftingSpinnerAdapter(ctx,
                CoreyAppUtils.getIntensitySpinnerData(ctx))
            spinnerIntensity.onItemSelectedListener = this

            recyclerViewExercises.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(ctx)
            exerciseAdapter = ExerciseAdapter(ctx)
            exerciseAdapter.setItemsMovable(true)
            // exerciseAdapter.setOnItemMoveListener(this);
            val callback = BaseItemTouchHelper(exerciseAdapter, true, BaseItemTouchHelper.DragAccess.VERTICAL)
            val touchHelper = ItemTouchHelper(callback)
            touchHelper.attachToRecyclerView(recyclerViewExercises)
            recyclerViewExercises.adapter = exerciseAdapter
        }

        if (isUpdateMode) {
            fillFields()
        }

        fragment_create_workout_btn_add_exercise.setOnClickListener {
            onClickAddExercise()
        }

        fragment_create_workout_btn_exp_col_general.setOnClickListener {
            onClickExpandCollapse()
        }
    }

    private fun onClickExpandCollapse() {

        if (cardState == CardState.EXPANDED) {
            ViewManager.collapse(generalCollapseContainer)
            imgBtnExpandCollapse.animate().rotation(-180f).start()
            cardState = CardState.COLLAPSED
        } else if (cardState == CardState.COLLAPSED) {
            ViewManager.expand(generalCollapseContainer)
            imgBtnExpandCollapse.animate().rotation(0f).start()
            cardState = CardState.EXPANDED
        }
    }

    private fun onClickAddExercise() {

        if (cardState == CardState.EXPANDED) {
            onClickExpandCollapse()
        }

        AddExercisesDialogFragment.newInstance()
                .setOnExerciseCreatedListener { exercise ->
                    exerciseAdapter.addEntityAtLast(exercise)
                }
                .show(childFragmentManager, "dialog-fragment-add-exercise")
    }

    private fun validateInput() {

        // Name -- Not empty
        val name = editName.text.toString()
        if (name.isEmpty()) {
            showSnackbar(getString(R.string.validation_empty_name))
            return
        }

        // Duration -- Not empty
        val strDuration = editDuration.text.toString()
        if (strDuration.isEmpty()) {
            showSnackbar(getString(R.string.validation_empty_duration))
            return
        }
        val duration = Integer.parseInt(strDuration)

        // Body Region -- Not first selected
        val brIdx = spinnerBodyRegion.selectedItemPosition
        if (brIdx <= 0) {
            showSnackbar(getString(R.string.validation_empty_body_region))
            return
        }

        // Intensity -- Not first selected
        val inIdx = spinnerIntensity.selectedItemPosition
        if (brIdx <= 0) {
            showSnackbar(getString(R.string.validation_empty_intensity))
            return
        }

        // Workout items -- Not empty
        val exercises = exerciseAdapter.data
        if (exercises.size == 0) {
            showSnackbar(getString(R.string.validation_empty_exercises))
            return
        }

        workout.setName(name)
        workout.duration = duration
        workout.bodyRegion = Workout.BodyRegion.values()[brIdx - 1]
        workout.intensity = Workout.Intensity.values()[inIdx - 1]
        workout.exercises = exercises

        back2Main()
    }

    private fun fillFields() {

        editName.setText(workout.displayableName)
        editDuration.setText(workout.duration.toString())
        val brIdx = workout.bodyRegion.ordinal + 1
        spinnerBodyRegion.setSelection(brIdx, true)
        val inIdx = workout.intensity.ordinal + 1
        spinnerIntensity.setSelection(inIdx, true)
        exerciseAdapter.data = workout.exercises
    }

    private fun back2Main() {
        val data = Intent()
                .putExtra(AppParams.INTENT_EXTRA_NEW_WORKOUT, workout)
                .putExtra(AppParams.INTENT_EXTRA_WORKOUT_UPDATED, isUpdateMode)
        activity?.setResult(Activity.RESULT_OK, data)
        activity?.supportFinishAfterTransition()
    }

    companion object {

        private const val ARG_WORKOUT = "arg_workout_edit"

        fun newInstance(workout: Workout?): CreateWorkoutFragment {
            val fragment = CreateWorkoutFragment()
            val args = Bundle()
            args.putParcelable(ARG_WORKOUT, workout)
            fragment.arguments = args
            return fragment
        }
    }
}
